tests/testthat/test-sptf-bulkdensity.R

```diff
@@ -1,10 +1,234 @@
+dt[, p143 := sptf_bd143(A_C_OF = A_C_OF,A_CLAY_MI = A_CLAY_MI,A_SAND_MI = A_SAND_MI,A_SILT_MI = A_SILT_MI)]
+dt[, p144 := sptf_bd144(A_C_OF = A_C_OF,A_CLAY_MI = A_CLAY_MI,A_SAND_MI = A_SAND_MI)]
+dt[, p145 := sptf_bd145(A_C_OF = A_C_OF,A_CLAY_MI = A_CLAY_MI,A_SAND_MI = A_SAND_MI,A_SILT_MI = A_SILT_MI)]
+dt[, p146 := sptf_bd146(A_C_OF = A_C_OF)]
+dt[, p147 := sptf_bd147(A_C_OF = A_C_OF,A_CLAY_MI = A_CLAY_MI,A_SAND_MI = A_SAND_MI,A_SILT_MI = A_SILT_MI,A_DEPTH = A_DEPTH)]
+dt[, p148 := sptf_bd148(A_SOM_LOI = A_SOM_LOI,A_CLAY_MI = A_CLAY_MI)]
+dt[, p149 := sptf_bd149(A_C_OF = A_C_OF)]
+dt[, p150 := sptf_bd150(A_C_OF = A_C_OF,A_DEPTH = A_DEPTH)]
+dt[, p151 := sptf_bd151(A_CLAY_MI = A_CLAY_MI,A_SAND_MI = A_SAND_MI,A_SILT_MI = A_SILT_MI)]
+dt[, p152 := sptf_bd152(A_SOM_LOI = A_SOM_LOI)]
+dt[, p153 := sptf_bd153(A_SOM_LOI = A_SOM_LOI, A_CLAY_MI = A_CLAY_MI)]
+dt[, p154 := sptf_bd154(A_C_OF = A_C_OF, A_PH_WA = A_PH_WA)]
+dt[, p155 := sptf_bd155(A_C_OF = A_C_OF, A_SILT_MI = A_SILT_MI)]
+dt[, p156 := sptf_bd156(A_C_OF = A_C_OF, A_CLAY_MI = A_CLAY_MI)]
+dt[, p157 := sptf_bd157(A_C_OF = A_C_OF, A_CLAY_MI = A_CLAY_MI,A_SILT_MI = A_SILT_MI)]
+dt[, p158 := sptf_bd158(A_C_OF = A_C_OF, A_CLAY_MI = A_CLAY_MI,A_SILT_MI = A_SILT_MI)]
+dt[, p159 := sptf_bd159(A_C_OF = A_C_OF, A_CLAY_MI = A_CLAY_MI)]
+dt[, p160 := sptf_bd160(A_SOM_LOI = A_SOM_LOI)]
+dt[, p161 := sptf_bd161(A_SOM_LOI = A_SOM_LOI)]
+dt[, p162 := sptf_bd162(A_C_OF = A_C_OF)]
+dt[, p163 := sptf_bd163(A_C_OF = A_C_OF)]
+dt[, p164 := sptf_bd164(A_SOM_LOI = A_SOM_LOI,A_CLAY_MI = A_CLAY_MI,A_SAND_MI = A_SAND_MI,A_SILT_MI = A_SILT_MI)]
+dt[, p165 := sptf_bd165(A_SOM_LOI = A_SOM_LOI)]
+dt[, p166 := sptf_bd166(A_C_OF = A_C_OF)]
+# ptf167 is soil density (i.e. soil weight without pore space), but not bulk density. Excluded.
+dt[, p167 := sptf_bd167(A_SOM_LOI = A_SOM_LOI,A_CLAY_MI = A_CLAY_MI)]
+dt[, p168 := sptf_bd168(A_SOM_LOI = A_SOM_LOI, A_CLAY_MI = A_CLAY_MI)]
+dt[, p169 := sptf_bd169(A_SOM_LOI = A_SOM_LOI,A_CLAY_MI = A_CLAY_MI)]
+dt[, p170 := sptf_bd170(A_SOM_LOI = A_SOM_LOI,A_CLAY_MI = A_CLAY_MI)]
+dt[, p171 := sptf_bd171(A_SOM_LOI = A_SOM_LOI,A_CLAY_MI = A_CLAY_MI)]
+dt[, p172 := sptf_bd172(A_SOM_LOI = A_SOM_LOI)]
+dt[, p173 := sptf_bd173(A_SOM_LOI = A_SOM_LOI)]
+dt[, p174 := sptf_bd174(A_SOM_LOI = A_SOM_LOI)]
+dt[, p175 := sptf_bd175(A_SOM_LOI = A_SOM_LOI)]
+dt[, p176 := sptf_bd176(A_SOM_LOI = A_SOM_LOI)]
+dt[, p177 := sptf_bd177(A_SOM_LOI = A_SOM_LOI,A_CLAY_MI = A_CLAY_MI,A_SILT_MI = A_SILT_MI,A_SAND_M50 = A_SAND_M50)]
+dt[, p178 := sptf_bd178(A_SOM_LOI = A_SOM_LOI,A_CLAY_MI = A_CLAY_MI)]
+dt[, p179 := sptf_bd179(A_SOM_LOI = A_SOM_LOI,A_C_OF = A_C_OF,A_CLAY_MI = A_CLAY_MI,A_SILT_MI = A_SILT_MI,A_SAND_MI = A_SAND_MI,A_PH_WA = A_PH_WA)]
+dt[, p180 := sptf_bd180(A_SOM_LOI = A_SOM_LOI,A_SAND_MI = A_SAND_MI,A_CLAY_MI = A_CLAY_MI)]
+dt[, p181 := sptf_bd181(A_SOM_LOI = A_SOM_LOI,A_CLAY_MI = A_CLAY_MI,A_SILT_MI = A_SILT_MI)]
```

Comment:
I discussed the unittests for bulkdensity briefly with SV.
It seems better after all to hardcode individual unit tests for each function. We can try to have GPT write the tests for us on Friday.
Extending the number of test cases from 2 to 5-10 may also be a good idea.

Reply:
I added tests for all functions. Now only 2 cases. We might later add some more